src/py_pecg/filtering_module.py

```diff
@@ -0,0 +1,564 @@
+            alpha (float): Variation of the cutoff frequency specification.
+
+        Returns:
+            b (TYPE): The resultant interpolation filter coefficients.
+        """
+        # filter specification (frequency and magnitude)
+        if alpha == 1:
+            M = np.array([r, r, 0, 0])
+            F = np.array([0, 1 / (2 * r), 1 / (2 * r), 0.5])
+        else:
+            Nband = int((np.floor(0.5 * r)))
+            M = np.concatenate(
+                (np.array([r, r]), np.zeros((1, 2 * Nband)).flatten())
+            )
+            a2r = alpha / 2 / r
+            F = np.zeros((1, 2 * Nband + 2)).flatten()
+            F[1] = a2r
+            k = 0
+            for i in range(2, 2 * Nband + 2, 2):
+                k = k + 1
+                F[i] = k / r - a2r
+                F[i + 1] = k / r + a2r
+            if F[2 * Nband + 1] > 0.5:
+                F[2 * Nband + 1] = 0.5
+            # design filter, cast to proper data type and convert it to a column vector
+            b = (
+                firls(2 * r * n + 1, 2 * F, M).astype(alpha.dtype).T
+            )  # firls returns a row vector
+            return b
+
+    # Argument Validation
+    if cutoff == 0:
+        cutoff = 0.5
+
+    if isinstance(idata, list):
+        raise TypeError("Input data must be a numpy array, not a list.")
+    if not isinstance(idata, np.ndarray):
+        raise TypeError("Input data must be a numpy array.")
+    if idata.size == 0:
+        raise ValueError("Input is an empty array.")
+    if idata.squeeze().ndim > 1:
+        raise ValueError("Input is a higher-dimensional array, not a matrix.")
+
+    if not isinstance(r, (int, np.integer)):
+        raise TypeError("r must be a finite integer scalar.")
+    if r < 1 or not np.isfinite(r):
+        raise ValueError("r must be >= 1 and finite.")
+
+    if not isinstance(n, (int, np.integer)):
+        raise TypeError("n must be a finite integer scalar.")
+    if n < 1 or not np.isfinite(n):
+        raise ValueError("n must be >= 1 and finite.")
+
+    if not isinstance(cutoff, (int, float)):
+        raise TypeError("cutoff must be a finite scalar.")
+    if cutoff <= 0 or cutoff > 1 or not np.isfinite(cutoff):
+        raise ValueError("cutoff must be > 0, <= 1, and finite.")
+
+    # Convert all input parameters to the same data type as the input signal
+    if idata.dtype == np.float32:
+        r = np.array(r, dtype=np.float32)
+        n = np.array(n, dtype=np.float32)
+        cutoff = np.array(cutoff, dtype=np.float32)
+    else:
+        r = np.array(r, dtype=np.float64)
+        n = np.array(n, dtype=np.float64)
+        cutoff = np.array(cutoff, dtype=np.float64)
+
+    # Determine if the input is a row vector
+    idataIsRow = idata.ndim == 1 and idata.shape[0] == idata.size
+    if idataIsRow:
+        xCol = np.reshape(idata, (-1, 1)).flatten()
+    else:
+        xCol = idata
```

Comment:
We can flatten without checking. If the array was already flat, it will do nothing.
```suggestion
     xCol = idata.flatten()
```